sdks/python/apache_beam/io/vcfio_test.py

```diff
@@ -0,0 +1,451 @@
+from source_test_utils import TestCaseWithTempDirCleanUp
+
+# Note: mixing \n and \r\n to verify both behaviors.
+_SAMPLE_HEADER_LINES = [
+    '##fileformat=VCFv4.2\n',
+    '##INFO=<ID=NS,Number=1,Type=Integer,Description="Number samples">\n',
+    '##INFO=<ID=AF,Number=A,Type=Float,Description="Allele Frequency">\n',
+    '##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">\r\n',
+    '##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Genotype Quality">\n',
+    '#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	Sample1	Sample2\r\n',
+]
+
+_SAMPLE_TEXT_LINES = [
+    '20	14370	.	G	A	29	PASS	AF=0.5	GT:GQ	0|0:48 1|0:48\n',
+    '20	17330	.	T	A	3	q10	AF=0.017	GT:GQ	0|0:49	0|1:3\n',
+    '20	1110696	.	A	G,T	67	PASS	AF=0.3,0.7	GT:GQ	1|2:21	2|1:2\n',
+    '20	1230237	.	T	.	47	PASS	.	GT:GQ	0|0:54	0|0:48\n',
+    '19	1234567	.	GTCT	G,GTACT	50	PASS	.	GT:GQ	0/1:35	0/2:17\n'
+    '20	1234	rs123	C	A,T	50	PASS	AF=0.5	GT:GQ	0/0:48	1/0:20\n',
+    '19	123	rs1234	GTC	.	40	q10;s50	NS=2	GT:GQ	1|0:48	0/1:.\n',
+    '19	12	.	C	<SYMBOLIC>	49	q10	AF=0.5	GT:GQ	0|1:45 .:.\n'
+]
+
+
+def get_full_file_path(file_name):
+  """Returns the full path of the specified ``file_name`` from ``data``."""
+  return os.path.join(
+      os.path.dirname(__file__), '..', 'testing', 'data', 'vcf', file_name)
+
+
+def get_full_dir():
+  """Returns the full path of the  ``data`` directory."""
+  return os.path.join(os.path.dirname(__file__), '..', 'testing', 'data', 'vcf')
+
+
+# Helper method for comparing variants.
+def _variant_comparator(v1, v2):
+  if v1.reference_name == v2.reference_name:
+    if v1.start == v2.start:
+      return cmp(v1.end, v2.end)
+    return cmp(v1.start, v2.start)
+  return cmp(v1.reference_name, v2.reference_name)
+
+
+# Helper method for verifying equal count on PCollection.
+def _count_equals_to(expected_count):
+  def _count_equal(actual_list):
+    actual_count = len(actual_list)
+    if expected_count != actual_count:
+      raise BeamAssertException(
+          'Expected %d not equal actual %d' % (expected_count, actual_count))
+  return _count_equal
+
+
+class VcfSourceTest(TestCaseWithTempDirCleanUp):
+
+  # Distribution should skip tests that need VCF files due to large size
+  VCF_FILE_DIR_MISSING = not os.path.exists(get_full_dir())
+
+  def _create_temp_vcf_file(self, lines, tmpdir=None):
+    return self._create_temp_file(suffix='.vcf', tmpdir=tmpdir, lines=lines)
+
+  def _read_records(self, file_or_pattern):
+    source = VcfSource(file_or_pattern)
+    range_tracker = source.get_range_tracker(None, None)
+    read_data = [record for record in source.read(range_tracker)]
+    return read_data
+
+  def _create_temp_file_and_read_records(self, lines):
+    return self._read_records(self._create_temp_vcf_file(lines))
+
+  def _assert_variants_equal(self, actual, expected):
+    self.assertEqual(
+        sorted(expected, cmp=_variant_comparator),
+        sorted(actual, cmp=_variant_comparator))
+
+  def _get_sample_variant_1(self):
```

Comment:
<!--new_thread; commit:63511ddb8caf9e5c12888ef8a4d1071e0ea6c168; resolved:0-->
What are the differences between methods get_sample_variant_1 _2 _3 etc. Can we use more descriptive names. At least we should clearly document the differences.

Reply:
Done.

Reply:
Doesn't seem done, the names are still the same.

Reply:
Since there are many special cases for each of the fields, and each field is processed independent of the others, it makes sense to have a few different variants that each have several special properties. The special features of each is described in a comment below the function name. Since each variant has several special features, there isn't really a good name that describes each...